### PR TITLE
refactor: remove asset balance calls to unconnected account areas

### PR DESCRIPTION
**Pull request by Prithpal-Sooriya**
## **Description**

One piece of the larger Unified Asset State Migration. We have a lot of components that indirectly use asset balanaces when they shouldn't ([see doc](https://github.com/MetaMask/metamask-extension/blob/temp/asset-selector-audit-report/trace-selectors/output/selector-usage-trace-assets-controllers.md)).

To help reduce the scope of changes required for the asset state migration we are cleaning up areas that shouldn't use asset state.

This should also help perf as these areas will stop receiving updates when balances or prices change.

This PR cleans up the unconnected account areas.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/40102?quickstart=1)

## **Changelog**

CHANGELOG entry: refactor: remove asset balance calls to unconnected account areas

## **Related issues**

Fixes: https://consensyssoftware.atlassian.net/browse/ASSETS-2666 https://consensyssoftware.atlassian.net/browse/ASSETS-2670

## **Manual testing steps**

N/A.

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots. 

---

> [!NOTE]
> **Medium Risk**
> Touches shared account/permissions selectors and their contract (adds `name`, changes underlying account source), which could affect any UI relying on `getOrderedConnectedAccountsForActiveTab` ordering/shape despite being a largely mechanical refactor.
> 
> **Overview**
> Removes balance-dependent account shaping from the *unconnected account* surfaces by switching `getOrderedConnectedAccountsForActiveTab` to use new balance-free selectors (`getMetaMaskAccountsWithoutBalance` / `getMetaMaskAccountsOrderedWithoutBalance`) and by having `UnconnectedAccountAlert`/`connected-accounts.container` pass connected accounts through directly.
> 
> Selector output is adjusted to include a top-level `name` sourced from `metadata.name`, and related PropTypes/stories/tests are updated (including dropping balance fields and loosening TS mocks) to match the new, balance-free account objects.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit c545c2c537edf92f7fded7fc093a44ddf46be622. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: ui/components/app/alerts/unconnected-account-alert/unconnected-account-alert.js [hunk 1]
```diff
@@ -40,16 +40,6 @@ const UnconnectedAccountAlert = () => {
...
   const internalAccounts = useSelector(getInternalAccounts);
-  // Temporary fix until https://github.com/MetaMask/metamask-extension/pull/21553
-  const internalAccountsMap = new Map(
-    internalAccounts.map((acc) => [acc.address, acc]),
-  );
-
-  const connectedAccountsWithName = connectedAccounts.map((account) => ({
-    ...account,
-    name: internalAccountsMap.get(account.address)?.metadata.name,
-  }));
```
Comment (Prithpal-Sooriya) [inline ui/components/app/alerts/unconnected-account-alert/unconnected-account-alert.js:51]: No need for this mapping step, instead the new selector will return an InternalAccount with a name property.

File: ui/components/app/connected-accounts-list/connected-accounts-list.component.js [hunk 2]
```diff
@@ -41,7 +41,6 @@ export default class ConnectedAccountsList extends PureComponent {
       PropTypes.shape({
         address: PropTypes.string.isRequired,
         name: PropTypes.string.isRequired,
-        lastActive: PropTypes.number,
```
Comment (Prithpal-Sooriya) [inline ui/components/app/connected-accounts-list/connected-accounts-list.component.js:44]: This property is never used, cleaning up

File: ui/components/app/connected-accounts-list/connected-accounts-list.stories.js [hunk 3]
```diff
@@ -25,14 +25,10 @@ export default {
       {
         name: 'This is a Really Long Account Name',
         address: '0x64a845a5b02460acf8a3d84503b0d68d028b4bb4',
-        index: 0,
-        balance: '0x176e5b6f173ebe66',
```
Comment (Prithpal-Sooriya) [inline ui/components/app/connected-accounts-list/connected-accounts-list.stories.js:29]: Cleaning up storybooks

File: ui/pages/connected-accounts/connected-accounts.container.js [hunk 4]
```diff
@@ -22,15 +22,7 @@ const mapStateToProps = (state) => {
...
   const internalAccounts = getInternalAccounts(state);
-  // Temporary fix until https://github.com/MetaMask/metamask-extension/pull/21553
-  const internalAccountsMap = new Map(
-    internalAccounts.map((acc) => [acc.address, acc]),
-  );
 
-  const connectedAccountsWithName = connectedAccounts.map((account) => ({
-    ...account,
-    name: internalAccountsMap.get(account.address)?.metadata.name,
-  }));
```
Comment (Prithpal-Sooriya) [inline ui/pages/connected-accounts/connected-accounts.container.js:33]: We do not need this complex mapping, instead the selector will return an internal account with a name prop.

File: ui/selectors/permissions.test.js [hunk 5]
```diff
@@ -427,8 +427,8 @@ describe('selectors', () => {
       expect(getOrderedConnectedAccountsForActiveTab(mockState)).toStrictEqual([
         {
           address: '0x8e5d75d60224ea0c33d0041e75de68b1c3cb6dd5',
-          balance: '0x0',
```
Comment (Prithpal-Sooriya) [inline ui/selectors/permissions.test.js:430]: removed balances

File: ui/selectors/permissions.test.js [hunk 6]
```diff
@@ -427,8 +427,8 @@ describe('selectors', () => {
...
           id: '07c2cfec-36c9-46c4-8115-3836d3ac9047',
+          name: 'Account 1',
```
Comment (Prithpal-Sooriya) [inline ui/selectors/permissions.test.js:431]: Added name property

File: ui/selectors/selectors.js [hunk 7]
```diff
@@ -691,14 +691,30 @@ export const getSelectedEvmInternalAccount = createSelector(
  *
  * @param keyrings - The array of keyrings.
  * @param accounts - The object containing the accounts.
- * @returns The array of internal accounts sorted by keyring.
+ * @returns {import('@metamask/keyring-internal-api').InternalAccount[]} The array of internal accounts sorted by keyring.
```
Comment (Prithpal-Sooriya) [inline ui/selectors/selectors.js:710]: JSDoc types for my sanity. I cannot work in JS 🤪

File: ui/selectors/selectors.js [hunk 8]
```diff
@@ -691,14 +691,30 @@ export const getSelectedEvmInternalAccount = createSelector(
...
  */
 export const getInternalAccountsSortedByKeyring = createSelector(
   getMetaMaskKeyrings,
-  getMetaMaskAccounts,
-  (keyrings, accounts) => {
+  getInternalAccounts,
```
Comment (Prithpal-Sooriya) [inline ui/selectors/selectors.js]: Important change, instead of using `getMetaMaskAccounts`, we can just use `getInternalAccounts`.

The shapes are almost identical. Main difference is that `getMetaMaskAccounts` returns an Object with accounts + balances, whereas `getInternalAccounts` returns an array of accounts.... (see below)

File: ui/selectors/selectors.js [hunk 9]
```diff
@@ -691,14 +691,30 @@ export const getSelectedEvmInternalAccount = createSelector(
...
     const thirdPartySnaps = 'thirdPartySnaps';
+
+    // Internal map of address to account for quick lookup
+    /** @type {Record<string, import('@metamask/keyring-internal-api').InternalAccount>} */
+    const accounts = accountsArr.reduce((map, account) => {
+      if (account?.address) {
+        map[account.address] = account;
+      }
+      return map;
+    }, {});
```
Comment (Prithpal-Sooriya) [inline ui/selectors/selectors.js]: ...this mapping basically makes the `getMetaMaskAccounts` but without balances :)

File: ui/selectors/selectors.js [hunk 10]
```diff
@@ -629,6 +629,21 @@ export const getMetaMaskAccounts = createChainIdSelector(
   },
 );
 
+export const getMetaMaskAccountsWithoutBalance = createSelector(
```
Comment (Prithpal-Sooriya) [inline ui/selectors/selectors.js:633]: yeah.... its hard since there are many files that use `getMetaMaskAccounts` (which that selector uses balances). Instead this acts as an interim until we refactor most/all `getMetaMaskAccounts` usages.

File: ui/selectors/selectors.js [hunk 11]
```diff
@@ -1025,6 +1052,24 @@ export const getMetaMaskAccountsOrdered = createSelector(
   },
 );
 
+/**
+ * Get internal accounts in an object format by address
+ *
+ * @returns {Record<string, import('@metamask/keyring-internal-api').InternalAccount>} An object of internal accounts by address
+ */
+export const getMetaMaskAccountsOrderedWithoutBalance = createSelector(
```
Comment (Prithpal-Sooriya) [inline ui/selectors/selectors.js:1061]: Same here, this is a nice drop in replacement for computing accounts without balances.

The number of nested selectors does not change, but at least we are slowly removing asset balances.

File: ui/selectors/selectors.js [hunk 12]
```diff
@@ -1025,6 +1052,24 @@ export const getMetaMaskAccountsOrdered = createSelector(
...
+  getMetaMaskAccountsWithoutBalance,
+  (internalAccounts, accounts) => {
+    return internalAccounts.map((internalAccount) => ({
+      ...internalAccount,
+      ...(internalAccount?.address
+        ? accounts[internalAccount.address] || {}
+        : {}),
+    }));
+  },
+);
```
[the remark this answers is not in the document]
Comment (Prithpal-Sooriya) [reply ui/selectors/selectors.js:1072]: You say that, but some existing tests are failing!
This is because the `getInternalAccountsSortedByKeyring` selector performs some keyring filtering.

TBH I do agree, but it would require a lot of extensive work to investigate and modify test fixtures.

File: ui/selectors/dapp.test.ts [hunk 13]
```diff
@@ -11,6 +11,10 @@ import {
 } from './selectors';
 import { getMultichainNetworkConfigurationsByChainId } from './multichain';
 
+// Mocked value for testing purposes only
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type MockedValue = any;
```
Comment (Prithpal-Sooriya) [inline ui/selectors/dapp.test.ts:16]: Only for testing purposes.

Because I've added JSDoc types, we now have tsc errors.
I would clean up the tests, but the actual implementation is still in JS so better safe to keep handling the edge cases

File: ui/selectors/selectors.js [hunk 14]
```diff
@@ -3356,14 +3407,22 @@ export const getOrderedConnectedAccountsForActiveTab = createDeepEqualSelector(
       },
     );
 
-    return orderedAccounts
+    /**
```
Comment (juanmigdr) [inline ui/selectors/selectors.js:3410]: Dont love the casting but if it is needed, it is what it is!

File: ui/selectors/selectors.js [hunk 15]
```diff
@@ -3343,9 +3388,15 @@ export function getUnconnectedAccounts(state, activeTab) {
...
   getOriginOfCurrentTab,
   (state) => state.metamask.permissionHistory,
-  getMetaMaskAccountsOrdered,
+  getMetaMaskAccountsOrderedWithoutBalance,
   getAllPermittedAccountsForCurrentTab,
-  (origin, permissionHistory, orderedAccounts, connectedAccounts) => {
+  (
+    origin,
+    permissionHistory,
+    /** @type {import('@metamask/keyring-internal-api').InternalAccount[]} */
```
Comment (juanmigdr) [inline ui/selectors/selectors.js:3396]: Dont love the casting but if it is needed, it is what it is!

File: ui/selectors/selectors.js [hunk 16]
```diff
@@ -1026,6 +1053,24 @@ export const getMetaMaskAccountsOrdered = createSelector(
...
+ * @returns {Record<string, import('@metamask/keyring-internal-api').InternalAccount>} An object of internal accounts by address
+ */
+export const getMetaMaskAccountsOrderedWithoutBalance = createSelector(
+  getInternalAccountsSortedByKeyring,
+  getMetaMaskAccountsWithoutBalance,
+  (internalAccounts, accounts) => {
+    return internalAccounts.map((internalAccount) => ({
+      ...internalAccount,
+      ...(internalAccount?.address
+        ? accounts[internalAccount.address] || {}
```
Comment (juanmigdr) [inline ui/selectors/selectors.js:1068]: accounts will override fields from internalAccounts and not the other way round, is this expected? Do we want to add a comment explaining it?

File: ui/selectors/selectors.js [hunk 17]
```diff
@@ -715,7 +737,10 @@ export const getInternalAccountsSortedByKeyring = createSelector(
     );
 
     // keep existing keyring order
-    return keyrings.reduce((internalAccounts, keyring) => {
+    /**
+     * @type {import('@metamask/keyring-internal-api').InternalAccount[]}
```
Comment (juanmigdr) [inline ui/selectors/selectors.js:741]: Same casting as below

**Review by juanmigdr — APPROVED** [review]
LGTM

**Review by bergarces — APPROVED** [review]

File: ui/selectors/selectors.js (hunk 14 above)
Comment (Prithpal-Sooriya) [reply ui/selectors/selectors.js:3410]: Yeah this is fkin JS files. Kill me 😭 

Hopefully we can move out more of these selectors into individual ts files.

File: ui/selectors/selectors.js (hunk 17 above)
Comment (Prithpal-Sooriya) [reply ui/selectors/selectors.js:741]: Treat this not as
```
const foo = {} as Foo
```

This is more like
```
const foo: Foo = {}
```

Intention is that if this file moves to using `// @@ts-check`, then this JS file will pick up type inferences

File: ui/selectors/selectors.js (hunk 16 above)
Comment (Prithpal-Sooriya) [reply ui/selectors/selectors.js:1068]: Great callout, this is intentional.

Realistically these point to the exact same data
- `internalAccounts` uses the internal accounts
- `accounts` uses the same internal accounts but restructured into a dictionary/map.
